src/combinatorics/simple_graph.lean

```diff
@@ -0,0 +1,176 @@
+/-
+Copyright (c) 2020 Aaron Anderson, Jalex Stark, Kyle Miller. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Aaron Anderson, Jalex Stark, Kyle Miller.
+-/
+import data.fintype.basic
+import data.sym2
+
+open finset
+
+/-!
+# Simple graphs
+
+This module defines simple graphs on a vertex type `V` as an
+irreflexive symmetric relation.
+
+There is a basic API for locally finite graphs and for graphs with
+finitely many vertices.
+
+- A locally finite graph is one with instances `∀ v, fintype (G.adj v)`.
+
+- Given instances `decidable_rel G.adj` and `fintype V`, then the graph
+is locally finite, too.
+
+## Implementation notes
+
+TODO: This is the simplest notion of an unoriented graph.  This should
+eventually fit into a more complete combinatorics hierarchy which
+includes multigraphs and directed graphs.  We begin with simple graphs
+in order to start learning what the combinatorics hierarchy should
+look like.
+
+TODO: Part of this would include defining, for example, subgraphs of a
+simple graph.
+
+-/
+
+universe u
+
+/--
+A simple graph is an irreflexive symmetric relation `adj` on a vertex type `V`.
+The relation describes which pairs of vertices are adjacent.
+There is exactly one edge for every pair of adjacent edges;
+see `simple_graph.E` for the corresponding type of edges.
+
+Note: The type of the relation is given as `V → set V` rather than
+`V → V → Prop` so that, given vertices `v` and `w`, then `w ∈ G.adj v`
+works as another way to write `G.adj v w`.  Otherwise Lean cannot find
+a `has_mem` instance.
+-/
+structure simple_graph (V : Type u) :=
+(adj : V → set V)
```

Comment:
I think I'm unconvinced by this. Why do we want the mem notation at all? I think it's best if the API commits to one or the other (i.e. always write `G.adj v w`, in which case this should be `V -> V -> Prop`, or always write `w \mem G.adj v`). Otherwise it's inviting hassle as people need to switch syntax.